### PR TITLE
Terrier weighting models expressed in Python

### PR DESCRIPTION
**Pull request by cmacdonald**

### CODE REVIEW

**Comment by cmacdonald** [comment]
(unit tests pass once terrier-python-helper is installed.)

File: pyterrier/batchretrieve.py [hunk 1]
```diff
@@ -570,6 +620,14 @@ def __setstate__(self, d):
         self.properties.update(d["properties"])
         for key,value in d["properties"].items():
             self.appSetup.setProperty(key, str(value))
+        #TODO consider the context state?
+
+    @staticmethod 
+    def from_dataset(dataset : Union[str,Dataset],
```
Comment (seanmacavaney) [inline pyterrier/batchretrieve.py:626]: looks like this belongs in a different PR?

**Review by seanmacavaney — COMMENTED** [review]
This is really cool!

Have you tested the performance difference between a pure java wmodel vs one of these? I'd expect there to be some impact, as there's overhead in jnius when switching between python and java. If it's noticeable, it might be worth mentioning this in the documentation.

Besides that, looks good to me.

File: pyterrier/batchretrieve.py (hunk 1 above)
Comment (cmacdonald) [reply pyterrier/batchretrieve.py:626]: I started from a from_dataset branch, but I think this bit is already merged.

**Comment by cmacdonald** [comment]
Efficiency timings on Vaswani for the record:
Baseline in 2.2120 seconds
Lambda in 81.0745 seconds

**Review by seanmacavaney — APPROVED** [review]
